cabal-install/Distribution/Client/CmdExec.hs

```diff
@@ -121,12 +145,58 @@ execAction execFlags extraArgs globalFlags = do
         (elaboratedShared buildCtx)
         buildStatus
 
-      -- TODO: discuss PATH munging with #hackage
+      -- Some dependencies may have executables. Let's put those on the PATH.
+      extraPaths <- pathAdditions verbosity buildCtx
 
       case extraArgs of
         exe:args -> runProgramInvocation
           verbosity
           (simpleProgramInvocation exe args)
             { progInvokeEnv = envOverrides
+            , progInvokePathEnv = extraPaths
             }
         [] -> die "Please specify an executable to run"
+
+pathAdditions :: Verbosity -> ProjectBuildContext -> IO [FilePath]
+pathAdditions verbosity ProjectBuildContext{..} = do
+  debug verbosity $ "Considering the following directories for inclusion in PATH:"
+  mapM_ (debug verbosity) paths
```

Comment:
Is there any particular reason to test if an executable exists on the physical file system, as opposed to unconditionally adding the binary directory to the path if the package in question defines some executables? You'd be able to save on a bunch of banging on the filesystem if you just go off the in-memory metadata.